### DISCUSSION TITLE
Prevent client from closing an application

### DISCUSSION THREAD

**Discussion by midiwidi** [Xpra-org/xpra#3874] [Q&A] [answered]
As the title says, is there a way of preventing a client from closing the application using the X on the GUI window? I couldn't find how to do that.
I'm starting the application in seamless mode with the following command
```
xpra start --bind-tcp=0.0.0.0:10001 --html=on --sharing=yes --pulseaudio=no --webcam=no --cursor=yes --file-transfer=off --resize-display=1920x1080 --start-new-commands=no --window-close=disconnect --session-name='My Application' --start='pipenv run python my_application.py'
```

The only idea I have for a work around is to start the application with --start-child and add --exit-with-children. Then monitor sessions and start the xpra server again if the session disappears.

**Comment by totaam** [comment, marked as answer]
Run your client with `--window-close=ignore` :
https://github.com/Xpra-org/xpra/blob/c3e5a3b9f9815852a48784b7f70b68a80a1cc8ab/fs/etc/xpra/conf.d/40_client.conf.in#L49-L54

**Comment by midiwidi** [comment]
I'm using the HTML5 client. The application (see below) is shared by multiple users and I want to prevent them from closing the application (red arrow).

![image](https://github.com/Xpra-org/xpra/assets/54610151/639f7b67-ab24-4683-a078-1318995c341f)

I commented `window-close = auto` and uncommented `window-close = ignore` in /etc/xpra/conf.d/40_client.conf but that didn't help. Users were still able to close the application window in the browser (red arrow). I've also tried to add `--window-close=ignore` to the `xpra start` options, but this didn't help either. Where do I need to apply that option?

**Comment by totaam** [comment]
> I commented window-close = auto and uncommented window-close = ignore in /etc/xpra/conf.d/40_client.conf but that didn't help. 

These options are for the python client.
> I'm using the HTML5 client

This should have been part of the original question.

This feature is not implemented in the http://github.com/Xpra-org/xpra-html5 client yet.
Feel free to open a github issue there, but bear in mind that I probably won't have the time to implement it...

**Reply by midiwidi** [reply]
Sorry, I didn't know. I thought it is a common client property.
But it also doesn't seem to work with the python client. Is that the one you mean?
![image](https://github.com/Xpra-org/xpra/assets/54610151/0a685a80-ab6b-4934-98f2-c74e5d07ecb5)
If I click the X on my application window which shows up after I launch the application on my Windows PC as shown in the screenshot, the application window gets closed. Not only on the Windows PC where I closed it (I expect that because I have specified `--window-close=disconnect` with the xpra start) but also in other clients which have it open (a Python client running on another Windows machine and on the Html5 client open in a browser on the same machine).

**Reply by totaam** [reply]
> I expect that because I have specified `--window-close=disconnect` with the `xpra start`

`window-close` is a client option, listed in the _client_ section of the man page, and in the _client_ configuration file.

**Reply by midiwidi** [reply]
Oh yes, you are right. This command option is left from when I just started to play around with xpra and had no experience.

I ran `xpra showconfig`
```
...
video-scaling                  = 'auto'
vsock-auth                     = 
webcam                         = 'auto'
window-close          (used)   = 'ignore'                          <class 'str'>
window-close         (default) = 'auto'                            <class 'str'>
window-icon                    = ''
windows                        = True
wm-name                        = 'Xpra'
...
```
and it shows that the window-close option is set to 'ignore'. 
Do you have an idea why the application window closes anyway? 

I had a look at the [server.log](https://github.com/Xpra-org/xpra/files/11599564/server.log)
Nothing came up when I closed the window in the Python client, but there are some permission errors. Not sure if that could cause the behavior. Do I need to run the xpra server as root to prevent those errors?

**Reply by totaam** [reply]
You need to use this configuration with the client.
Perhaps using the launcher does not honour this specific option.
Try running the client from the command line, ie: `xpra attach ssh://user@host/ --window-close=ignore` (you should be able to remove the option)

**Reply by midiwidi** [reply]
Oh, now I understand. The whole time I was making the settings in /etc/xpra/conf.d/40_client.conf on the server machine. I thought the server is communicating the settings to the clients when the connect. But one has to make the window-close setting on the client sides for each client. Then it works as expected. 
Wouldn't it make more sense if the server had control over this and could prevent them from closing the application?

**Reply by totaam** [reply]
Making this setting a server-driven one would be much more complicated.

**Reply by midiwidi** [reply]
Understood
